src/ray/core_worker/store_provider/store_provider.h

```diff
@@ -1,34 +1,71 @@
 #ifndef RAY_CORE_WORKER_STORE_PROVIDER_H
 #define RAY_CORE_WORKER_STORE_PROVIDER_H
 
-#include "ray/common/buffer.h"
 #include "ray/common/id.h"
 #include "ray/common/status.h"
 #include "ray/core_worker/common.h"
 
+namespace arrow {
+  class Buffer;
+namespace py {
+struct SerializedPyObject;
+}
+}
+
 namespace ray {
 
-/// Binary representation of ray object.
+/// RayObjects consist of data and metadata. Metadata is always passed
+/// in as a buffer because it is small, but data can be handled 
+/// differently by implementations in order to avoid copying large data.
 class RayObject {
  public:
-  /// Create a ray object instance.
+  /// Create a ray object.
   ///
-  /// \param[in] data Data of the ray object.
   /// \param[in] metadata Metadata of the ray object.
-  RayObject(const std::shared_ptr<Buffer> &data, const std::shared_ptr<Buffer> &metadata)
-      : data_(data), metadata_(metadata) {}
+  RayObject(const std::shared_ptr<arrow::Buffer> &metadata) 
+      : metadata_(metadata) {}
+
+  const std::shared_ptr<arrow::Buffer> &Metadata() const { return metadata_; };
+
+  virtual size_t DataSize();
+
+  virtual Status WriteDataTo(std::shared_ptr<arrow::Buffer> buffer);
```

Comment:
This class also needs a virtual destructor;

Also: Should we make `WriteDataTo` a purely virtual method (with `= 0`) to make sure the base class cannot be used?

Reply:
Good idea.